src/Microsoft.AspNetCore.OData/Extensions/HttpRequestExtensions.cs

```diff
@@ -451,6 +452,57 @@ public static void DeleteRequestContainer(this HttpRequest request, bool dispose
             }
         }
 
+        /// <summary>
+        /// Checks whether the request is a POST targeted at a resource path ending in /$query.
+        /// </summary>
+        /// <param name="request">The request.</param>
+        /// <param name="oDataPath">The OData path.</param>
+        /// <returns>true if the request path has $query segment.</returns>
+        internal static bool IsQueryRequest(this HttpRequest request, string oDataPath)
+        {
+            return request.Method.Equals(HttpMethods.Post) && 
+                oDataPath?.TrimEnd('/').EndsWith('/' + ODataRouteConstants.QuerySegment, StringComparison.OrdinalIgnoreCase) == true;
+        }
+
+        /// <summary>
+        /// Transforms a POST request targeted at a resource path ending in $query into a GET request. 
+        /// The query options are parsed from the request body and appended to the request URL.
+        /// </summary>
+        /// <param name="request">The request.</param>
+        /// <param name="requestContainerFactory">Request container factory.</param>
+        internal static void TransformQueryRequest(this HttpRequest request, Func<IServiceProvider> requestContainerFactory)
+        {
+            if (requestContainerFactory == null)
+            {
+                throw Error.ArgumentNull("requestContainerFactory");
+            }
+
+            IServiceProvider requestContainer = requestContainerFactory();
+
+            // Fetch parsers available in the request container for parsing the query options in the request body
+            IEnumerable<IODataQueryOptionsParser> queryOptionsParsers = requestContainer.GetRequiredService<IEnumerable<IODataQueryOptionsParser>>();
+            IODataQueryOptionsParser queryOptionsParser = queryOptionsParsers.FirstOrDefault(
+                d => d.MediaTypeMapping.TryMatchMediaType(request) > 0);
+
+            if (queryOptionsParser == null)
+            {
+                throw new ODataException(string.Format(
+                    CultureInfo.InvariantCulture, 
+                    SRResources.CannotFindParserForRequestMediaType,
+                    request.ContentType));
+            }
+
+            string queryString = queryOptionsParser.Parse(request.Body);
+
+            // Request path starts with a /
+            string requestPath = request.Path.Value;
+            requestPath = requestPath.Substring(0, requestPath.LastIndexOf('/' + ODataRouteConstants.QuerySegment, StringComparison.OrdinalIgnoreCase));
+
+            request.Path = new PathString(requestPath);
+            request.QueryString = new QueryString(queryString);
+            request.Method = HttpMethods.Get;
```

Comment:
@habbes I'm not exact sure I understand what you're asking here. However, just to note that this feature rides on the content-type, based on OASIS OData standard [here](http://docs.oasis-open.org/odata/odata/v4.01/csprd05/part2-url-conventions/odata-v4.01-csprd05-part2-url-conventions.html#sec_PassingQueryOptionsintheRequestBody). In addition, the request from the client will be a POST

Reply:
I think he's asking if we should remove headers describing the request body (i.e., content-type and content-length) from the transformed query request, since they don't apply to the transformed GET request. Might be nice to clean them up, but not sure it's really a lot of value, and we would have to enumerate the full list of such headers to strip out.  As long as nothing downstream complains about these non-relevant headers, I think I'd just leave them.

---
In reply to: [437210686](https://github.com/OData/WebApi/pull/2191#discussion_r437210686) [](ancestors = 437210686)